packages/database/src/cachers/EntityHierarchyCacher.js

```diff
@@ -5,17 +5,126 @@
 import { reduceToDictionary } from '@tupaia/utils';
 import { ORG_UNIT_ENTITY_TYPES } from '../modelClasses/Entity';
 
+const REBUILD_DEBOUNCE_TIME = 1000; // wait 1 second after changes before rebuilding, to avoid double-up
+
 export class EntityHierarchyCacher {
   constructor(models) {
     this.models = models;
-    this.generationsVisited = new Set();
+    this.changeHandlerCancellers = [];
+    this.resetScheduledHierarchyRebuild();
+  }
+
+  listenForChanges() {
+    this.changeHandlerCancellers[0] = this.models.entity.addChangeHandler(this.handleEntityChange);
+    this.changeHandlerCancellers[1] = this.models.entityRelation.addChangeHandler(
+      this.handleEntityRelationChange,
+    );
+  }
+
+  stopListeningForChanges() {
+    this.changeHandlerCancellers.forEach(c => c());
+    this.changeHandlerCancellers = [];
+  }
+
+  handleEntityChange = async ({ record_id: entityId }) => {
+    // if entity was deleted or created, or parent_id has changed, we need to delete subtrees and
+    // rebuild all hierarchies
+    const hierarchies = await this.models.entityHierarchy.all();
+    const hierarchyTasks = hierarchies.map(async ({ id: hierarchyId }) => {
+      await this.deleteSubtree(hierarchyId, entityId);
+      return this.scheduleHierarchyForRebuild(hierarchyId);
+    });
+    await Promise.all(hierarchyTasks);
+  };
+
+  handleEntityRelationChange = async ({ record }) => {
+    // delete subtree from parent_id onwards within the associated hierarchy, and kick off a new build
+    // of any projects associated with changed hierarchies
+    const { entity_hierarchy_id: hierarchyId, parent_id: parentId } = record;
+    await this.deleteSubtree(hierarchyId, parentId);
+    return this.scheduleHierarchyForRebuild(hierarchyId);
+  };
+
+  resetScheduledHierarchyRebuild() {
+    this.hierarchiesForRebuild = new Set();
+    this.scheduledRebuildPromise = null;
+    this.resolveScheduledRebuild = null;
+    this.scheduledRebuildTimeout = null;
```

Comment:
These actually do get instantiated in the constructor by the `this.resetScheduledHierarchyRebuild()`. I had them listed out in the constructor explicitly, but then realised it would be more DRY to do it this way (e.g. if we wanted to change the `hierarchiesForRebuild` to an array rather than a Set, we would risk missing one if they were in two places). With that in mind, do you still suggest I should list them explicitly up the top, and just have them set to `undefined` initially? That's probably the way to go huh...

Reply:
I guess so... But you know what, let's not worry about it because we haven't been consistent the whole time. And perhaps it doesn't really matter in a JavaScript context...